File: apps/api-gql/internal/delivery/gql/mappers/custom_overlays.go
package mappers

import (
	"github.com/twirapp/twir/apps/api-gql/internal/delivery/gql/gqlmodel"
	"github.com/twirapp/twir/apps/api-gql/internal/entity"
)

func CustomOverlayToGql(e entity.CustomOverlay) gqlmodel.CustomOverlay {
	layers := make([]gqlmodel.CustomOverlayLayer, len(e.Layers))
	for i, layer := range e.Layers {
		layers[i] = gqlmodel.CustomOverlayLayer{
			ID:                      layer.ID,
			Type:                    gqlmodel.CustomOverlayLayerType(layer.Type),
			OverlayID:               layer.OverlayID,
			Width:                   layer.Width,
			Height:                  layer.Height,
			CreatedAt:               layer.CreatedAt,
			UpdatedAt:               layer.UpdatedAt,
			PeriodicallyRefetchData: layer.PeriodicallyRefetchData,
			TransformString:         layer.TransformString,
		}

		if layer.SettingsHtmlHtml != nil && layer.SettingsHtmlCss != nil && layer.SettingsHtmlJs != nil {
			layers[i].Settings = gqlmodel.CustomOverlayLayerSettingsHTML{
				EmptyHolder:         nil,
				HTML:                *layer.SettingsHtmlHtml,
				CSS:                 *layer.SettingsHtmlCss,
				Js:                  *layer.SettingsHtmlJs,
				PollSecondsInterval: *layer.SettingsHtmlDataPoll,
			}
		}

		if layer.SettingsImageURL != nil {
			layers[i].Settings = gqlmodel.CustomOverlayLayerSettingsImage{
				EmptyHolder: nil,
				URL:         *layer.SettingsImageURL,
			}
		}
	}

	return gqlmodel.CustomOverlay{
		ID:        e.ID,
		ChannelID: e.ChannelID,
		Name:      e.Name,
		CreatedAt: e.CreatedAt,
		UpdatedAt: e.UpdatedAt,
		Width:     e.Width,
		Height:    e.Height,
		Layers:    layers,
	}
}

File: frontend/dashboard/src/api/custom-overlays.ts
import { useQuery } from '@urql/vue'
import { createGlobalState } from '@vueuse/core'

import { useMutation } from '@/composables/use-mutation.js'
import { graphql } from '@/gql/gql.js'

export const useCustomOverlaysApi = createGlobalState(() => {
	const cacheKey = 'customOverlays'

	const useData = () => useQuery({
		query: graphql(`
			query CustomOverlays {
				customOverlays {
					id
					name
					createdAt
					updatedAt
					width
					height
					layers {
						id
						type
						overlayId
						width
						height
						createdAt
						updatedAt
						periodicallyRefetchData
						transformString
						settings {
							... on CustomOverlayLayerSettingsHTML {
								html
								css
								js
								pollSecondsInterval
							}
							... on CustomOverlayLayerSettingsImage {
								url
							}
						}
					}
				}
			}
		`),
		context: {
			additionalTypenames: [cacheKey],
		},
		variables: {},
	})

	const useDelete = () => useMutation(
		graphql(`
			mutation DeleteCustomOverlay($id: UUID!) {
				customOverlaysDelete(id: $id)
			}
		`),
		[cacheKey],
	)

	const useCreate = () => useMutation(
		graphql(`
			mutation CreateCustomOverlay($createInput: CustomOverlayCreateInput!) {
				customOverlaysCreate(input: $createInput) {
					id
				}
			}
		`),
		[cacheKey],
	)

	const useUpdate = () => useMutation(
		graphql(`
			mutation UpdateCustomOverlay($updateInput: CustomOverlayUpdateInput!) {
				customOverlaysUpdate(input: $updateInput) {
					id
				}
			}
		`),
		[cacheKey],
	)

	const useParseVariablesInHtml = () => useMutation(graphql(`
		mutation ParseVariablesInHtml($text: String!) {
			customOverlaysParseHtml(input: { text: $text})
		}
	`))

	return {
		useData,
		useDelete,
		useCreate,
		useUpdate,
		useParseVariablesInHtml,
	}
})
